tools/gfx/d3d12/render-d3d12.cpp

```diff
@@ -841,93 +638,1214 @@ class D3D12Renderer : public GraphicsAPIRenderer
+            }
+
+            virtual SLANG_NO_THROW void SLANG_MCALL setIndexBuffer(
+                IBufferResource* buffer,
+                Format indexFormat,
+                UInt offset = 0) override
+            {
+                m_boundIndexBuffer = (BufferResourceImpl*)buffer;
+                m_boundIndexFormat = D3DUtil::getMapFormat(indexFormat);
+                m_boundIndexOffset = UINT(offset);
+            }
+
+            void prepareDraw()
+            {
+                auto pipelineState = m_currentPipeline.Ptr();
+                if (!pipelineState || (pipelineState->desc.type != PipelineType::Graphics))
+                {
+                    assert(!"No graphics pipeline state set");
+                    return;
+                }
+
+                // Submit - setting for graphics
+                {
+                    GraphicsSubmitter submitter(m_d3dCmdList);
+                    _bindRenderState(static_cast<PipelineStateImpl*>(pipelineState), &submitter);
+                }
+
+                m_d3dCmdList->IASetPrimitiveTopology(m_primitiveTopology);
+
+                // Set up vertex buffer views
+                {
+                    int numVertexViews = 0;
+                    D3D12_VERTEX_BUFFER_VIEW vertexViews[16];
+                    for (Index i = 0; i < m_boundVertexBuffers.getCount(); i++)
+                    {
+                        const BoundVertexBuffer& boundVertexBuffer = m_boundVertexBuffers[i];
+                        BufferResourceImpl* buffer = boundVertexBuffer.m_buffer;
+                        if (buffer)
+                        {
+                            D3D12_VERTEX_BUFFER_VIEW& vertexView = vertexViews[numVertexViews++];
+                            vertexView.BufferLocation =
+                                buffer->m_resource.getResource()->GetGPUVirtualAddress() +
+                                boundVertexBuffer.m_offset;
+                            vertexView.SizeInBytes =
+                                UINT(buffer->getDesc()->sizeInBytes - boundVertexBuffer.m_offset);
+                            vertexView.StrideInBytes = UINT(boundVertexBuffer.m_stride);
+                        }
+                    }
+                    m_d3dCmdList->IASetVertexBuffers(0, numVertexViews, vertexViews);
+                }
+                // Set up index buffer
+                if (m_boundIndexBuffer)
+                {
+                    D3D12_INDEX_BUFFER_VIEW indexBufferView;
+                    indexBufferView.BufferLocation =
+                        m_boundIndexBuffer->m_resource.getResource()->GetGPUVirtualAddress() +
+                        m_boundIndexOffset;
+                    indexBufferView.SizeInBytes =
+                        UINT(m_boundIndexBuffer->getDesc()->sizeInBytes - m_boundIndexOffset);
+                    indexBufferView.Format = m_boundIndexFormat;
+
+                    m_d3dCmdList->IASetIndexBuffer(&indexBufferView);
+                }
```

Comment:
Not relevant to this PR, but we may be close to a point where we can move some of the logic that is in `prepareDraw()` to just be in the relevant state-setting function. E.g., there seems to be no reason why we can't call `IASetIndexBuffer` as part of the `setIndexBuffer` implementation at this point.

The main sticking point that remains is dealing with the way that the root shader object plus the pipeline state combine to produce the final state, but that seems almost inevitable to have to deal with.

Reply:
That's right, should clean up in a future pass.